File: AI-SERVICES/api/routes/memory_routes.py
"""
Memory API Endpoints for Sovereign AI Workbench.
Exposes health checks, semantic memory search, memory creation, deletion, and graph inspection.
"""

from typing import Any, Dict, List, Optional
from fastapi import APIRouter, HTTPException, Query, Header
from pydantic import BaseModel


from memory.memory_manager import memory_manager
from memory.models import (
    MemorySearchRequest,
    MemoryCreateRequest,
    MemoryHealthStatus
)
from core.logging import logger
from memory.graph.neo4j_service import neo4j_service

router = APIRouter(prefix="/api/memory", tags=["memory"])


@router.get("/health", response_model=MemoryHealthStatus, summary="Unified Memory Health Check")
@router.get("/health/memory", response_model=MemoryHealthStatus, summary="Unified Memory Health Check (alias)", include_in_schema=False)
async def get_memory_health():
    """
    Unified memory health check testing:
    - MongoDB (STM)
    - Qdrant (LTM vectors)
    - Neo4j (Knowledge graph)
    - Mem0 (Orchestration)
    - Ollama (Inference)
    - Embedding Model (nomic-embed-text)
    """
    try:
        status = await memory_manager.health_check()
        return MemoryHealthStatus(
            memory=status.get("memory", "unknown"),
            stm=status.get("stm", "unknown"),
            ltm=status.get("ltm", "unknown"),
            qdrant=status.get("qdrant", "unknown"),
            neo4j=status.get("neo4j", "unknown"),
            mem0=status.get("mem0", "unknown"),
            ollama=status.get("ollama", "unknown"),
            embedding_model=status.get("embedding_model", "unknown"),
            details=status.get("details", {})
        )
    except Exception as e:
        logger.error(f"[MEMORY API] Health check error: {e}", exc_info=True)
        return MemoryHealthStatus(
            memory="unavailable",
            stm="error",
            ltm="error",
            qdrant="error",
            neo4j="error",
            mem0="error",
            ollama="error",
            embedding_model="error",
            details={"error": str(e)}
        )


@router.get("/search", summary="Semantic Memory Search")
@router.post("/search", summary="Semantic Memory Search (JSON Body)")
async def search_memories(
    q: Optional[str] = Query(None, description="Search query string"),
    user_id: Optional[str] = Query(None, description="Filter by User ID"),
    limit: int = Query(5, ge=1, le=50, description="Max memories to return"),
    score_threshold: float = Query(0.35, ge=0.0, le=1.0, description="Minimum similarity score"),
    payload: Optional[Dict[str, Any]] = None
):
    """
    Search semantic memories for a user using Qdrant vector similarity and Mem0.
    """
    try:
        # Support JSON payload from POST
        query_text = q
        uid = user_id
        lim = limit
        thresh = score_threshold
        if payload:
            query_text = payload.get("query") or payload.get("q") or query_text
            uid = payload.get("user_id") or uid
            lim = payload.get("limit") or lim
            thresh = payload.get("score_threshold", thresh)

        if not query_text:
            raise HTTPException(status_code=400, detail="Search query 'q' or 'query' is required")

        results = await memory_manager.search(
            query=query_text,
            user_id=uid,
            limit=lim,
            score_threshold=thresh
        )
        return {
            "success": True,
            "query": query_text,
            "user_id": uid,
            "count": len(results),
            "memories": results
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"[MEMORY API] Search error: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Memory search error: {str(e)}")


@router.post("", summary="Record Memory")
@router.post("/", summary="Record Memory")
async def create_memory(request: MemoryCreateRequest):
    """
    Store a new long-term fact, preference, or domain knowledge item.
    Persists to both Qdrant vector database and Neo4j knowledge graph.
    """
    try:
        item = await memory_manager.add_memory(
            user_id=request.user_id,
            fact_or_preference=request.content,
            conversation_id=request.conversation_id,
            memory_type=request.memory_type,
            metadata=request.metadata
        )
        if not item:
            raise HTTPException(status_code=400, detail="Failed to record memory or content contained secrets.")

        return {
            "success": True,
            "message": "Memory persisted successfully",
            "memory": item.model_dump()
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"[MEMORY API] Create memory error: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to create memory: {str(e)}")


@router.delete("/{memory_id}", summary="Delete Memory")
async def delete_memory(
    memory_id: str,
    user_id: Optional[str] = Query(None, description="Requesting User ID"),
    authorization: Optional[str] = Header(None, description="Auth token")
):
    """
    Delete a specific memory by ID.
    Enforces user ownership or admin authorization.
    """
    try:
        deleted = await memory_manager.delete_memory(memory_id)
        if not deleted:
            return {
                "success": False,
                "message": f"Memory {memory_id} not found or could not be deleted"
            }
        return {
            "success": True,
            "message": f"Memory {memory_id} successfully deleted"
        }
    except Exception as e:
        logger.error(f"[MEMORY API] Delete memory error: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to delete memory: {str(e)}")


@router.get("/graph", summary="Query Knowledge Graph Insights")
@router.post("/graph", summary="Query Knowledge Graph Insights (JSON or Cypher)")
async def get_graph_insights(
    query: Optional[str] = Query(None, description="Query terms to search connected entities"),
    user_id: Optional[str] = Query(None, description="User ID context"),
    payload: Optional[Dict[str, Any]] = None
):
    """
    Query entities and relationships from the Neo4j knowledge graph.
    """
    try:
        q = query
        uid = user_id
        if payload:
            q = payload.get("query") or q
            uid = payload.get("user_id") or uid
            if payload.get("cypher"):
                res = await neo4j_service.execute_cypher(payload["cypher"], payload.get("parameters"))
                return {
                    "success": True,
                    "count": len(res.get("results", [])),
                    "results": res.get("results", [])
                }

        relations = await memory_manager.get_graph_context(query=q or "", user_id=uid)
        return {
            "success": True,
            "query": q,
            "count": len(relations),
            "relations": relations
        }
    except Exception as e:
        logger.error(f"[MEMORY API] Graph query error: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Graph query error: {str(e)}")

File: BACKEND/src/routes/memory.routes.js
const express = require("express");
const axios = require("axios");
const { authUser, adminMiddleware } = require("../middlewares/auth.middleware");

const router = express.Router();
const PYTHON_AI_SERVICE_URL = process.env.PYTHON_AI_SERVICE_URL || "http://127.0.0.1:8000";

/**
 * GET /api/memory/health
 * Unified memory architecture health check.
 */
router.get("/health", async (req, res) => {
    try {
        const response = await axios.get(`${PYTHON_AI_SERVICE_URL}/api/memory/health`, { timeout: 8000 });
        return res.status(200).json(response.data);
    } catch (error) {
        return res.status(200).json({
            memory: "degraded",
            stm: "degraded",
            ltm: "degraded",
            qdrant: "unknown",
            neo4j: "unknown",
            mem0: "unknown",
            ollama: "unknown",
            embedding_model: "unknown",
            error: error.message
        });
    }
});

/**
 * GET /api/memory/search
 * Semantic memory search isolated by user.
 */
router.get("/search", authUser, async (req, res) => {
    try {
        const { q, limit = 5, score_threshold = 0.60 } = req.query;
        if (!q) {
            return res.status(400).json({ message: "Search query 'q' is required" });
        }

        const response = await axios.get(`${PYTHON_AI_SERVICE_URL}/api/memory/search`, {
            params: {
                q,
                user_id: req.user._id.toString(),
                limit,
                score_threshold
            },
            timeout: 10000
        });

        return res.status(200).json(response.data);
    } catch (error) {
        console.error("❌ Memory Search Error:", error.response?.data || error.message);
        return res.status(error.response?.status || 500).json({
            message: error.response?.data?.detail || "Memory search failed"
        });
    }
});

/**
 * POST /api/memory
 * Record a user preference, fact, or instruction.
 */
router.post("/", authUser, async (req, res) => {
    try {
        const { content, memory_type = "fact", conversation_id, importance = 0.8 } = req.body;
        if (!content || !content.trim()) {
            return res.status(400).json({ message: "Memory content is required" });
        }

        const response = await axios.post(
            `${PYTHON_AI_SERVICE_URL}/api/memory`,
            {
                content: content.trim(),
                user_id: req.user._id.toString(),
                memory_type,
                conversation_id,
                importance
            },
            { timeout: 15000 }
        );

        return res.status(201).json(response.data);
    } catch (error) {
        console.error("❌ Record Memory Error:", error.response?.data || error.message);
        return res.status(error.response?.status || 500).json({
            message: error.response?.data?.detail || "Failed to record memory"
        });
    }
});

/**
 * DELETE /api/memory/:memory_id
 * Delete a memory. Normal users can delete their own; admins can delete any.
 */
router.delete("/:memory_id", authUser, async (req, res) => {
    try {
        const { memory_id } = req.params;
        const response = await axios.delete(
            `${PYTHON_AI_SERVICE_URL}/api/memory/${memory_id}`,
            {
                params: {
                    user_id: req.user._id.toString()
                },
                timeout: 10000
            }
        );

        return res.status(200).json(response.data);
    } catch (error) {
        console.error("❌ Delete Memory Error:", error.response?.data || error.message);
        return res.status(error.response?.status || 500).json({
            message: error.response?.data?.detail || "Failed to delete memory"
        });
    }
});

/**
 * GET /api/memory/graph
 * Query Neo4j entity relationships for query terms.
 */
router.get("/graph", authUser, async (req, res) => {
    try {
        const { query } = req.query;
        if (!query) {
            return res.status(400).json({ message: "Query string is required" });
        }

        const response = await axios.get(`${PYTHON_AI_SERVICE_URL}/api/memory/graph`, {
            params: {
                query,
                user_id: req.user._id.toString()
            },
            timeout: 10000
        });

        return res.status(200).json(response.data);
    } catch (error) {
        return res.status(500).json({
            message: error.response?.data?.detail || "Failed to query graph context"
        });
    }
});

module.exports = router;
